lib/vfs/src/host_fs.rs

```diff
@@ -0,0 +1,648 @@
+            host_path,
+            flags,
+        }
+    }
+
+    pub fn metadata(&self) -> fs::Metadata {
+        self.inner.metadata().unwrap()
+    }
+}
+
+impl Read for HostFile {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        self.inner.read(buf)
+    }
+    fn read_to_end(&mut self, buf: &mut Vec<u8>) -> io::Result<usize> {
+        self.inner.read_to_end(buf)
+    }
+    fn read_to_string(&mut self, buf: &mut String) -> io::Result<usize> {
+        self.inner.read_to_string(buf)
+    }
+    fn read_exact(&mut self, buf: &mut [u8]) -> io::Result<()> {
+        self.inner.read_exact(buf)
+    }
+}
+impl Seek for HostFile {
+    fn seek(&mut self, pos: io::SeekFrom) -> io::Result<u64> {
+        self.inner.seek(pos)
+    }
+}
+impl Write for HostFile {
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.inner.write(buf)
+    }
+    fn flush(&mut self) -> io::Result<()> {
+        self.inner.flush()
+    }
+    fn write_all(&mut self, buf: &[u8]) -> io::Result<()> {
+        self.inner.write_all(buf)
+    }
+    fn write_fmt(&mut self, fmt: ::std::fmt::Arguments) -> io::Result<()> {
+        self.inner.write_fmt(fmt)
+    }
+}
+
+#[cfg_attr(feature = "enable-serde", typetag::serde)]
+impl VirtualFile for HostFile {
+    fn last_accessed(&self) -> u64 {
+        self.metadata()
+            .accessed()
+            .ok()
+            .and_then(|ct| ct.duration_since(SystemTime::UNIX_EPOCH).ok())
+            .map(|ct| ct.as_nanos() as u64)
+            .unwrap_or(0)
+    }
+
+    fn last_modified(&self) -> u64 {
+        self.metadata()
+            .modified()
+            .ok()
+            .and_then(|ct| ct.duration_since(SystemTime::UNIX_EPOCH).ok())
+            .map(|ct| ct.as_nanos() as u64)
+            .unwrap_or(0)
+    }
+
+    fn created_time(&self) -> u64 {
+        self.metadata()
+            .created()
+            .ok()
+            .and_then(|ct| ct.duration_since(SystemTime::UNIX_EPOCH).ok())
+            .map(|ct| ct.as_nanos() as u64)
+            .unwrap_or(0)
+    }
+
+    fn size(&self) -> u64 {
+        self.metadata().len()
+    }
+
+    fn set_len(&mut self, new_size: u64) -> Result<(), FsError> {
+        fs::File::set_len(&self.inner, new_size).map_err(Into::into)
+    }
+
+    fn unlink(&mut self) -> Result<(), FsError> {
+        std::fs::remove_file(&self.host_path).map_err(Into::into)
+    }
+    fn sync_to_disk(&self) -> Result<(), FsError> {
+        self.inner.sync_all().map_err(Into::into)
+    }
+
+    fn bytes_available(&self) -> Result<usize, FsError> {
+        // unwrap is safe because of get_raw_fd implementation
+        let host_fd = self.get_raw_fd().unwrap();
+
+        host_file_bytes_available(host_fd)
+    }
+
+    #[cfg(unix)]
+    fn get_raw_fd(&self) -> Option<i32> {
+        use std::os::unix::io::AsRawFd;
+        Some(self.inner.as_raw_fd())
+    }
+    #[cfg(not(unix))]
+    fn get_raw_fd(&self) -> Option<i32> {
+        unimplemented!(
```

Comment:
I believe we may be able to use [`AaRawHandle`](https://doc.rust-lang.org/std/os/windows/io/trait.AsRawHandle.html) and [`RawHandle`](https://doc.rust-lang.org/std/os/windows/io/type.RawHandle.html) here. It's the size of a pointer (`type RawHandle = HANDLE;` and `type HANDLE = *mut c_void`), so we should update `i32` that represents a file descriptor.

Ideally, we should come with a type to represent a file descriptor:

```rust
type FileDescriptor = usize;

#[cfg(unix)]
impl From<RawFd> for FileDescriptor { … }

#[cfg(windows)]
impl From<RawHandle> for FileDescriptor { … }
```

Reply:
Draft here, https://github.com/wasmerio/wasmer/pull/2528